client/my-sites/controller.js

```diff
@@ -293,12 +298,16 @@ function isPathAllowedForDIFMInProgressSite( path, slug, domains, contextParams
 function onSelectedSiteAvailable( context ) {
 	const state = context.store.getState();
 	const selectedSite = getSelectedSite( state );
+	// Check the possibly expired plan to ensure it's a trial plan.
+	const maybeExpiredPlanSlug = selectedSite?.plan?.product_slug;
 	// Use getSitePlanSlug() as it ignores expired plans.
 	const currentPlanSlug = getSitePlanSlug( state, selectedSite.ID );
 
-	// If we had a trial plan, and the user doesn't have an active paid plan, redirect to fullpage trial expired page.
+	// If we had a trial plan, and the user doesn't have a paid plan (active or expired),
+	// redirect to full-page trial expired page.
 	if (
 		wasTrialSite( state, selectedSite.ID ) &&
+		PLAN_ECOMMERCE_TRIAL_MONTHLY === maybeExpiredPlanSlug &&
```

Comment:
Quick note in passing, as I will only be able to review tomorrow: this change means that we'll only perform this redirect while the expired plan is still linked to the site. I believe that will only be a for a week post-expiration, so we may need to cater for a null/undefined value for `maybeExpiredPlanSlug`.
```suggestion
		[ PLAN_ECOMMERCE_TRIAL_MONTHLY, null, undefined ].includes( maybeExpiredPlanSlug ) &&
```

I think we'll need additional changes to support the behaviour for views that are more than 7 days after expiration.

Reply:
Oh interesting, I will test this with a longer expiration date and see what happens!

Reply:
Updating the expiration date to be further back than one week doesn't seem to change the results with this PR, but it seems possible that updating the expiration date in store admin does not affect the value in state and so we wouldn't see accurate results until an actual week has passed.

If we check for null/undefined there I think we'll end up in the same situation, where someone with a long expired non-Woo Express Trial plan will see the trial expiration screen.

I'll keep digging and see if I can figure out how to reliably find the site's last active plan.

Reply:
I think the main issue is that expired subscriptions are only removed ~7 days after expiration via a background process, so it may be more appropriate to test this with an older site where you had the free trial, purchased a Woo Express plan (ideally a monthly plan!), and then had the plan expire some time ago. (I am sure I have some sites that meet this criteria!) I think we would also have a similar issue for sites where you started the free trial and never upgraded, but your free trial expired and was then removed. AFAIK, we can't distinguish between these two values in Calypso. We may need to expose our tracking for sites that upgraded from the free trial via the Sites endpoint, much like we did for the `was_ecommerce_trial` flag.

If we do that, it's probably worth using a generic `was_upgraded_from_trial` flag to cover both the eCommerce/Woo Express and Business trials, which would require changes in Jetpack and WordPress.com, specifically in the following locations:
* Add logic to compute the value in `SAL_Site` - https://github.com/Automattic/jetpack/blob/a715bbe0a47ba6d72bd592c1594fdf6d5857e751/projects/plugins/jetpack/sal/class.json-api-site-base.php#L491-L503
* Add the field to the JSON endpoint, and add logic to compute it from `$this->site`
   - https://github.com/Automattic/jetpack/blob/a715bbe0a47ba6d72bd592c1594fdf6d5857e751/projects/plugins/jetpack/json-endpoints/class.wpcom-json-api-get-site-v1-2-endpoint.php#L76
   - https://github.com/Automattic/jetpack/blob/a715bbe0a47ba6d72bd592c1594fdf6d5857e751/projects/plugins/jetpack/json-endpoints/class.wpcom-json-api-get-site-endpoint.php

We'd then have this available to the UI to know when the user ever made a purchase after the trial, which is all we really need to determine

Reply:
Thanks for pointing me in the right direction @daledupreez ! I started on the Jetpack side of things here: https://github.com/Automattic/jetpack/pull/34597

It looks like we have a `has-upgraded-from-ecommerce-trial` sticker already, but it only gets applied when someone upgrades from the Woo Express Trial -> Ecommerce plan. I think moving that sticker assignment into the `activate` method of the main `Store_Product_Bundle` class would solve the problem for sites that upgrade to *any* plan. Then we can leverage it in the Jetpack SAL and create a new selector for it in Calypso. Thoughts?

Reply:
It looks like we also have a `has-upgraded-from-business-trial` sticker for upgrades from the hosting and migration trials as well.

Reply:
Just to close the loop we discussed this in p1702403127249889-slack-C0Q664T29 and I've updated the PRs to check for the presence of any of the `has-upgraded-from-${trial}-trial` stickers so I don't need to move the sticker logic around. 🥳